README.md
# Visual Studio Code - iTunes & Apple Music Player

Control iTunes or Apple Music From Within With Visual Studio Code.

This is a fork of the original from [PsykoSoldi3r](https://github.com/PsykoSoldi3r/vscode-itunes), as the project appears to have been abandoned. This version adds settings, custom formatting options, and rich tooltip support.

<img src="https://raw.githubusercontent.com/dnedry2/vscode-itunes/master/images/Tooltip-V.png" width="256"> <img src="https://raw.githubusercontent.com/dnedry2/vscode-itunes/master/images/Tooltip-H.png" width="256">

## Features

* Display Current Playing Track
    * Track
    * Artist
    * Podcast Title
* Rich tooltips with album art
* Media controls
    * Previous Track
    * Play/Pause Track
    * Next Track
    * Open iTunes
    * Mute/Unmute Volume
    * Set Repeat Mode: One, All or Off
    * Toggle Shuffle
    * Like & Dislike 
    * Add To Library

## Language Support

Following languages are supported. Can be set in the settings of Visual Studio Code
* English (en), default
* Chinese (zh)
* Dutch (nl)

Your language not supported? Please submit a new issue with the translated AppleScripts: `repeatSetAll`, `repeatSetOff`, `repeatSetOne`, `shuffleOff`, `shuffleOn`

## Requirements

* System: macOS
* Application: iTunes & Apple Music


CHANGELOG.md
# Changelog

## v0.19.0
### Added
* Media control buttons to tooltip
* Vertical tooltip layout

## v0.18.0
### Added
* Rich tooltip with album art

## v0.17.0
### Forked

### Added
* Visibility settings for all controls
* Custom title formatting

## v0.16.1
### Fixed
* Wrong icon of button 'add the current playing track to library'.

## v0.16.0
### Added
* Able to add the current playing track to your iTunes/Apple Music library.

## v0.15.0
### Added
* Added a maximun length of artist and title text. This can be changed in the settings of Visual Studio Code. The default is 64. To turn it off, set the length to 0.

## v0.14.2
### Fixed
* Fixed issues with Dutch Apple Scripts

## v0.14.1
### Fixed
* Fixed issues with Chinese Apple Scripts

## v0.14.0
### Added
* Like & Dislike buttons

## v0.13.0
### Added
* Multilanguage Dutch and Chinese (beta)

## v0.12.0
### Added
* Updated play icon

## v0.11.0
### Added
* Apple Music Support - Catalina ([#21](https://github.com/PsykoSoldi3r/vscode-itunes/issues/21))

## v0.10.0
### Fixed
* Songs with Double quotes does not get displayed ([#13](https://github.com/PsykoSoldi3r/vscode-itunes/issues/13))
* Does not work with Remote Development (VS Code Insiders)([#17](https://github.com/PsykoSoldi3r/vscode-itunes/issues/17))
* No control display ([#18](https://github.com/PsykoSoldi3r/vscode-itunes/issues/18))

## v0.9.1
### Fixed
* Double quotes in title of a track won't be displayed

## v0.9.0
### Added
* Display podcast title ([#11](https://github.com/PsykoSoldi3r/vscode-itunes/issues/11))

## v0.8.3
### Bugfixes
* Don't display (non breaking) errors ([#6](https://github.com/PsykoSoldi3r/vscode-itunes/issues/6))

## v0.8.1
### Changes
* Updated [README.md](https://github.com/PsykoSoldi3r/vscode-itunes/blob/master/README.md)

## v0.8.0
### Added
* Turn shuffle on or off
* Tooltips

## v0.7.3
### Security
* Updated dependecies with security vulnerability

## v0.7.2
### Security
* Updated dependecies with security vulnerability

## v0.7.1
### Added
* Command itunes.volume

## v0.7.0
### Added
* Mute/unmute sound when clicking on the volume indiciator button

## v0.6.0
### Changes
* Hide text when iTunes is stopped and not opened

## v0.5.0
### Added
* `itunes.updateInterval` for status checking ( default value: 5000 )

## v0.4.0
### Added
* Repeat Button: All, One or Off

## v0.3.0
### Changes
* Reposition media controls

## v0.2.0
### Added
* Navigate to next track
* Navigate to previous track
* Open iTunes

## v0.1.2
### Added
* Show play/pause button when iTunes is offline or idle

## v0.1.1
### Bugfixes
* Bugfix: Start iTunes by starting `vscode`

## v0.1.0

* Initial release of itunes-vscode


src/Config.ts
import { workspace, WorkspaceConfiguration } from "vscode";

export default class Config {

    public static Instance: Config;
    private workspace: WorkspaceConfiguration;

    constructor(){
        this.workspace = workspace.getConfiguration('itunes');
        Config.Instance = this;
    }

    getStatusCheckInterval(){
        const interval = this.workspace.get("updateInterval", 5000 );
        if( interval < 1000 ){
            return 1000;
        }

        return interval;
    }

    getLanguageOverride(){
        const languageOverride = this.workspace.get("languageOverride", "en" );
        return languageOverride;
    }

    getSkipTrackDislike(): boolean {
        const skipTrackByTrackDislike = this.workspace.get("skipTrackByTrackDislike", true);
        return skipTrackByTrackDislike;
    }

    getTitleStringLimit(): number {
        const titleStringLimit = this.workspace.get("titleStringLimit", 0);
        return titleStringLimit;
    }

    getShowMediaControls(): boolean {
        const showControls = this.workspace.get("showMediaControls", true);
        return showControls;
    }

    getShowAdd(): boolean {
        const showAdd = this.workspace.get("showAdd", true);
        return showAdd;
    }

    getShowTrack(): boolean {
        const showTrack = this.workspace.get("showTrack", true);
        return showTrack;
    }

    getTrackFormat(): string {
        const trackFormat = this.workspace.get("trackFormat", "{name} - {artist} — {album}");
        return trackFormat;
    }

    getShowLoopShuffle(): boolean {
        const showLS = this.workspace.get("showLoopShuffle", true);
        return showLS;
    }

    getShowMute(): boolean {
        const showMute = this.workspace.get("showMute", true);
        return showMute;
    }

    getShowLikeDislike(): boolean {
        const showLD = this.workspace.get("showLikeDislike", true);
        return showLD;
    }

    getVerticalLayout(): boolean {
        const useV = this.workspace.get("useVerticalTooltip", true);
        return useV;
    }
}

src/iTunes.ts
import ITrack from "./ITrack";
import * as util from "util";
import { exec } from "child_process";
import * as applescript from "applescript";
import * as path from "path";
import { window } from "vscode";
import vscode = require("vscode");
import { release, userInfo } from "os";
import { env } from "process";
import * as os from "os";
import * as fs from "fs";
import Config from "./Config";

export default class iTunes {

    private _application = "itunes";

    constructor(){
        const version = release();
        const majorVersion = version.split(".")[0] || 0;
        if( majorVersion >= 19 ){
            this._application = "music";
        }
    }

    public getAppState(): Promise<{}> {
        return this.executeScript("appState");
    }

    public getCurrentTrack(): Promise<{}> {
        return this.executeScript("currentTrack", false);
    }

    public play(): void {
        this.executeScript("play");
    }

    public pause(): void {
        this.executeScript("pause");
    }

    public nextTrack(): void {
        this.executeScript("nextTrack");
    }

    public previousTrack(): void {
        this.executeScript("previousTrack");
    }

    public open(): void {
        this.executeScript("open");
    }

    public unmute(): void {
        this.executeScript("unmute");
    }

    public mute(): void {
        this.executeScript("mute");
    }

    public like(): void {
        this.executeScript("likeTrack");
    }

    public dislikeTrack(): void {
        this.executeScript("dislikeTrack");
    }

    public dislikeTrackAndSkip(): void {
        this.executeScript("dislikeTrackAndSkip");
    }
    
    public shuffle( enable: boolean ): void {
        if( enable === true ) {
            this.executeScript("shuffleOn");
        } else {
            this.executeScript("shuffleOff");
        }
    }

    public addTrack(): void {
        this.executeScript("addToLibrary");
    }

    public setRepeat( repeat: string ): void {
        this.executeScript(`repeatSet${repeat}`)
            .then( ( result ) => {
                if( result == null ){
                    window.showErrorMessage("Visual Studio Code hasn't access to Accessibilty of your macOS. Please enable at System Preferences -> Security & Privacy");
                }
            })
            .catch( () => {
                window.showErrorMessage("Visual Studio Code hasn't access to Accessibilty of your macOS. Please enable at System Preferences -> Security & Privacy");
            });
    }

    public saveArt(): Promise<{}> {
        return this.executeScript("saveArt", false);
    }

    private getScript( filename, app: string = "music", language: string = "en" ): string {
        const file = path.resolve(__dirname, `../../scripts/${app}/${language}/${filename}.applescript`);
        if ( fs.existsSync( file ) ) {
            return file;
        }

        return path.resolve(__dirname, `../../scripts/${app}/en/${filename}.applescript`);
    }

    private executeScript( filename: string, isJson: boolean = true ) : Promise<{}>{
        return new Promise( ( resolve, reject ) => {
            const language = Config.Instance.getLanguageOverride();
            const script = this.getScript( filename, this._application, language );
            applescript.execFile( script, ( err, result ) => {
                if( err ){
                    reject( err );
                }

                if( result != null ){
                    try {
                        if( isJson === true ) {
                            const parsedString = JSON.parse( result.toString() );
                            resolve( parsedString );
                        } else {
                            resolve( result );
                        }
                    }catch( exception ){ }
                }else{
                    resolve({});
                }
            });
        });
    }
}


src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import Player from './Player';
import Config from './Config';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
    const config = new Config();
    const player = new Player();

    const commandPlay = vscode.commands.registerCommand('itunes.play', () => {
        player.play();
        player.forceUpdate();
    });

    const commandPause = vscode.commands.registerCommand('itunes.pause', () => {
        player.pause();
        player.forceUpdate();
    });

    const commandNextTrack = vscode.commands.registerCommand("itunes.nextTrack", () => {
        player.nextTrack();
        player.forceUpdate();
    });

    const commandPreviousTrack = vscode.commands.registerCommand("itunes.previousTrack", () => {
        player.previousTrack();
        player.forceUpdate();
    });

    const commandOpen = vscode.commands.registerCommand("itunes.open", () => {
        player.open();
        player.forceUpdate();
    });

    const commandVolume = vscode.commands.registerCommand("itunes.volume", () => {
        player.volume();
        player.forceUpdate();
    });

    const commandRepeatOne = vscode.commands.registerCommand("itunes.repeat.set.one", () => {
        player.setRepeat("One");
        player.forceUpdate();
    });

    const commandRepeatOff = vscode.commands.registerCommand("itunes.repeat.set.off", () => {
        player.setRepeat("Off");
        player.forceUpdate();
    });

    const commandRepeatAll = vscode.commands.registerCommand("itunes.repeat.set.all", () => {
        player.setRepeat("All");
        player.forceUpdate();
    });

    const commandShuffleOn = vscode.commands.registerCommand("itunes.shuffle.on", () => {
        player.shuffleOn();
        player.forceUpdate();
    });

    const commandShuffleOff = vscode.commands.registerCommand("itunes.shuffle.off", () => {
        player.shuffleOff();
        player.forceUpdate();
    });

    const likeTrack = vscode.commands.registerCommand("itunes.likeTrack", () => {
        player.likeTrack();
        player.forceUpdate();
    });

    const dislikeTrack = vscode.commands.registerCommand("itunes.dislikeTrack", () => {
        player.dislikeTrack();
        player.forceUpdate();
    });

    const addTrack = vscode.commands.registerCommand("itunes.addTrack", () => {
        player.addTrack();
        player.forceUpdate();
    });

    context.subscriptions.push( player );
    context.subscriptions.push( commandPlay );
    context.subscriptions.push( commandPause );
    context.subscriptions.push( commandNextTrack );
    context.subscriptions.push( commandPreviousTrack );
    context.subscriptions.push( commandOpen );
    context.subscriptions.push( commandRepeatOne );
    context.subscriptions.push( commandRepeatOff );
    context.subscriptions.push( commandRepeatAll );
    context.subscriptions.push( commandVolume );
    context.subscriptions.push( commandShuffleOn );
    context.subscriptions.push( commandShuffleOff );
    context.subscriptions.push( likeTrack );
    context.subscriptions.push( dislikeTrack );
    context.subscriptions.push( addTrack );
}

// this method is called when your extension is deactivated
export function deactivate() {
    Player.Instance.dispose();
}


src/Player.ts
import { window, StatusBarAlignment, StatusBarItem, workspace, MarkdownString, } from "vscode";
import iTunes from "./iTunes";
import ITrack, { MediaType } from "./ITrack";
import Config from "./Config";
import TrackFactory from "./Factories/TrackFactory";
import { print } from "util";
import { SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS } from "constants";
import { writeFileSync, unlinkSync } from "fs";
import { tmpdir } from "os";
import { config } from "process";

export default class Player {
    public static Instance: Player;

    private iTunes: iTunes;
    private titleBarItem: StatusBarItem = null;
    private playerButton: StatusBarItem = null;
    private previousTrackButton: StatusBarItem = null;
    private nextTrackButton: StatusBarItem = null;
    private stateButton: StatusBarItem = null;
    private updateInterval: NodeJS.Timer;
    private repeatButton: StatusBarItem = null;
    private shuffleButton: StatusBarItem = null;
    private likeButton: StatusBarItem = null;
    private dislikeButton: StatusBarItem = null;
    private addToLibrayButton: StatusBarItem = null;

    private lastSongName: string = "";
    private artworkPath:  string = "";
    private artworkSaved: boolean = false;
    private defaultArtwork: string = "data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPHN2ZyBkYXRhLW5hbWU9IkxheWVyIDEiIHZpZXdCb3g9IjAgMCA2NCA2NCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPGRlZnM+CjxzdHlsZT4uY2xzLTF7ZmlsbDojZmZmO30uY2xzLTJ7ZmlsbDp1cmwoI2EpO308L3N0eWxlPgo8bGluZWFyR3JhZGllbnQgaWQ9ImEiIHgxPSIxMC4xIiB4Mj0iNTguNjciIHkxPSI0Ny4wNyIgeTI9IjEzLjY1IiBncmFkaWVudFVuaXRzPSJ1c2VyU3BhY2VPblVzZSI+CjxzdG9wIHN0b3AtY29sb3I9IiNkZTVkNzQiIG9mZnNldD0iMCIvPgo8c3RvcCBzdG9wLWNvbG9yPSIjN2UzMWYzIiBvZmZzZXQ9Ii41Ii8+CjxzdG9wIHN0b3AtY29sb3I9IiMxMmUyZWEiIG9mZnNldD0iMSIvPgo8L2xpbmVhckdyYWRpZW50Pgo8L2RlZnM+CjxyZWN0IGNsYXNzPSJjbHMtMSIgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0IiByeD0iNiIvPgo8cGF0aCBjbGFzcz0iY2xzLTIiIGQ9Ik0zMiwyLjUzQTI5LjQ3LDI5LjQ3LDAsMSwwLDYxLjQ3LDMyLDI5LjQ2LDI5LjQ2LDAsMCwwLDMyLDIuNTNabTAsMzcuMjVBNy43OCw3Ljc4LDAsMSwxLDM5Ljc4LDMyLDcuNzgsNy43OCwwLDAsMSwzMiwzOS43OFoiLz4KPC9zdmc+";

    private statusBarPositionOffset: number = 10;

    private spacer = "<img src=\"data:image/svg+xml;base64,PHN2Zz48L3N2Zz4=\" width=\"350\" height=\"0\">";

    private richTooltipTemplateH: string = "[<img src=\"$artwork\" width=\"64\" align=\"left\"/>](command:itunes.open \"Open iTunes\")<span><b>&nbsp;&nbsp;$name</b></span></br><span>&nbsp;&nbsp;$(organization)&nbsp;$artist</span></br><span>&nbsp;&nbsp;$(list-flat)&nbsp;$album</span></br>";
    private richTooltipControls: string  = "<h1>[$(thumbsup)](command:itunes.likeTrack \"Like track\")[$(thumbsdown)](command:itunes.dislikeTrack \"Dislike track\")&nbsp;&nbsp;&nbsp;&nbsp;[$(plus)](command:itunes.addTrack \"Add to library\")&nbsp;&nbsp;&nbsp;&nbsp;[$($shuffleIcon)](command:itunes.shuffle.$shuffleCmd)&nbsp;[$(chevron-left)](command:itunes.previousTrack \"Previous\")[$($playIcon)](command:itunes.$playCmd \"Play / Pause\")[$(chevron-right)](command:itunes.nextTrack \"Next\")&nbsp;[$($repeatIcon)](command:itunes.repeat.set.$repeatCmd)&nbsp;&nbsp;&nbsp;&nbsp;[$(mute)](command:itunes.volume \"Mute\")</h1>";

    private richTooltipTemplateV: string = "[<img src=\"$artwork\" width=\"320\"/>](command:itunes.open \"Open iTunes\")</br><h3>$name</h3><span>$(organization)&nbsp;$artist</span></br><span>$(list-flat)&nbsp;$album</span></br>";


    constructor() {
        Player.Instance = this;

        this.onUpdate = this.onUpdate.bind( this );

        this.iTunes = new iTunes();

        this.createStatusBarItem();
        this.updateInterval = setInterval( this.onUpdate, Config.Instance.getStatusCheckInterval() );
        this.updateStatusBarItem();

        // Reset on settings change
        workspace.onDidChangeConfiguration(e => {
            Config.Instance = new Config()
            this.updateInterval = setInterval( this.onUpdate, Config.Instance.getStatusCheckInterval() );
            this.hideMediaControls();
            this.showMediaControls();
        });
    }

    private onUpdate(){
        this.updateStatusBarItem();
    }

    private createStatusBarItem(){
        this.previousTrackButton = window.createStatusBarItem( StatusBarAlignment.Left, 14 + this.statusBarPositionOffset );
        this.previousTrackButton.text = "$(chevron-left)";
        this.previousTrackButton.command = "itunes.previousTrack";
        this.previousTrackButton.tooltip = "Play Previous Track";

        this.playerButton = window.createStatusBarItem( StatusBarAlignment.Left, 13 + this.statusBarPositionOffset );

        this.nextTrackButton = window.createStatusBarItem( StatusBarAlignment.Left, 12 + this.statusBarPositionOffset );
        this.nextTrackButton.text = "$(chevron-right)";
        this.nextTrackButton.command = "itunes.nextTrack";
        this.nextTrackButton.tooltip = "Play Next Track";

        this.addToLibrayButton = window.createStatusBarItem( StatusBarAlignment.Left, 11 + this.statusBarPositionOffset );
        this.addToLibrayButton.text = "$(plus)";
        this.addToLibrayButton.command = "itunes.addTrack";
        this.addToLibrayButton.tooltip = "Add current playing to track to your iTunes/Apple Music master library."

        this.titleBarItem = window.createStatusBarItem( StatusBarAlignment.Left, 10 + this.statusBarPositionOffset );
        this.titleBarItem.command = "itunes.open";
        this.titleBarItem.tooltip = "Show iTunes";
        this.titleBarItem.show();

        this.repeatButton = window.createStatusBarItem( StatusBarAlignment.Left, 9 + this.statusBarPositionOffset );
        this.shuffleButton = window.createStatusBarItem( StatusBarAlignment.Left, 8 + this.statusBarPositionOffset );

        this.stateButton = window.createStatusBarItem( StatusBarAlignment.Left, 7 + this.statusBarPositionOffset );
        this.stateButton.text = "$(mute)";
        this.stateButton.command = "itunes.volume";

        this.likeButton = window.createStatusBarItem( StatusBarAlignment.Left, 6 + this.statusBarPositionOffset );
        this.likeButton.text = "$(thumbsup)";
        this.likeButton.command = "itunes.likeTrack";

        this.dislikeButton = window.createStatusBarItem( StatusBarAlignment.Left, 5 + this.statusBarPositionOffset );
        this.dislikeButton.text = "$(thumbsdown)";
        this.dislikeButton.command = "itunes.dislikeTrack";
    }

    private updateStatusBarItem(){
        this.iTunes.getAppState()
            .then( ( app: any ) => {
                if( app.appState === "running" ){
                    this.iTunes.getCurrentTrack()
                        .then( ( track: any ) => {
                            const currentTrack: ITrack = TrackFactory.create( track );

                            if (currentTrack.artist != null && currentTrack.name != null) {
                                // Save new artwork if song has changed
                                if (this.lastSongName != currentTrack.name) {
                                    this.lastSongName = currentTrack.name;
                                    this.artworkSaved = false;
                                    
                                    this.iTunes.saveArt()
                                        .then( ( bin: any ) => {
                                            const path = tmpdir() + "/vscode-itunes-artwork-" + currentTrack.name + ".jpg";

                                            if (this.artworkPath != "") {
                                                unlinkSync(this.artworkPath);
                                                this.artworkPath = "";
                                            }

                                            writeFileSync(path, bin, { flag: 'w', });

                                            this.artworkPath = path;
                                            this.artworkSaved = true;

                                            this.forceUpdate();
                                        })
                                        .catch( ( err ) => {
                                            console.log( err );
                                        });
                                }

                                this.updateStatusText( currentTrack.artist, currentTrack.name, currentTrack.album, currentTrack.kind, currentTrack.state === "playing", currentTrack.repeat_song, currentTrack.shuffle === "true");

                                this.titleBarItem.show();

                                if (Config.Instance.getShowMute())
                                    this.stateButton.show();
                            }

                            if( currentTrack.volume >= 1 ){
                                this.stateButton.text = "$(unmute)";
                                this.stateButton.tooltip = "Mute Volume";
                            }else{
                                this.stateButton.text = "$(mute)";
                                this.stateButton.tooltip = "Unmute Volume";
                            }

                            switch( currentTrack.state ){
                                case "playing" :
                                    this.playerButton.text = "$(primitive-square)";
                                    this.playerButton.command = "itunes.pause";
                                    this.playerButton.tooltip = "Pause Track";
                                    break;

                                case "paused" :
                                case "stopped" :
                                    this.playerButton.text = "$(play)";
                                    this.playerButton.command = "itunes.play";
                                    this.playerButton.tooltip = "Play Track";
                                    break;
                            }

                            switch( currentTrack.repeat_song ){
                                case "all":
                                    this.repeatButton.text = "$(sync) All";
                                    this.repeatButton.command = "itunes.repeat.set.one";
                                    break;

                                case "one":
                                    this.repeatButton.text = "$(sync) One";
                                    this.repeatButton.command = "itunes.repeat.set.off";
                                    break;

                                case "off":
                                    this.repeatButton.text = "$(sync) Off";
                                    this.repeatButton.command = "itunes.repeat.set.all";
                                    break;
                            }

                            if( currentTrack.shuffle === "true" ) {
                                this.shuffleButton.text = "$(git-compare) On";
                                this.shuffleButton.command = "itunes.shuffle.off";
                                this.shuffleButton.tooltip = "Turn Shuffle Off";
                            } else {
                                this.shuffleButton.text = "$(git-compare) Off";
                                this.shuffleButton.command = "itunes.shuffle.on";
                                this.shuffleButton.tooltip = "Turn Shuffle On";
                            }

                            if( currentTrack.loved === true ) {
                                this.likeButton.text = "$(thumbsup) $(check)";
                            } else {
                                this.likeButton.text = "$(thumbsup)";
                            }

                            if( currentTrack.disliked === true ) {
                                this.dislikeButton.text = "$(thumbsdown) $(check)";
                            } else {
                                this.dislikeButton.text = "$(thumbsdown)";
                            }

                            this.showMediaControls();
                        })
                        .catch( ( err ) => {
                            console.log( err );
                        });
                }else{
                    this.hideMediaControls();
                }
            });
    }

    private updateStatusText( artist: string, name: string, album: string, kind: MediaType, playing: boolean, repeat: string, shuffle: boolean) {
        const useVertLayout = Config.Instance.getVerticalLayout();
        const titleStringLimit = Config.Instance.getTitleStringLimit();
        const title = Config.Instance.getTrackFormat().replace("{name}", name)
                                                      .replace("{artist}", artist)
                                                      .replace("{album}", album);

        let displayedTitle = title;

        if( titleStringLimit > 0 ) {
            displayedTitle = title.substring(0, titleStringLimit);
            if( displayedTitle.length < title.length ) {
                displayedTitle += "…";
            }
        }

        let artwork = this.defaultArtwork;
        if (this.artworkSaved) {
            artwork = this.artworkPath;
        }

        /*
        let scrubAlbum = album;
        let scrubArtist = artist.substring(0, name.length);

        if (album.length + 6 > name.length)
            scrubAlbum = album.substring(0, name.length - 6) + "…";

        if (artist.length + 6 > name.length)
            scrubArtist = artist.substring(0, name.length - 6) + "…";
*/

        let template = useVertLayout ? this.richTooltipTemplateV : this.richTooltipTemplateH;

        let richTooltip = new MarkdownString(template.replace("$name",    name)
                                                     .replace("$artist",  artist)
                                                     .replace("$album",   album)
                                                     .replace("$artwork", artwork), true);

        richTooltip.isTrusted   = true;
        richTooltip.supportHtml = true;

        let repeatCmd  = "";
        let repeatIcon = "";
        switch (repeat) {
            case "all":
                repeatCmd = "one \"Repeat: All\"";
                repeatIcon = "debug-restart-frame"
                break;

            case "one":
                repeatCmd = "off \"Repeat: One\"";
                repeatIcon = "debug-restart"
                break;

            case "off":
                repeatCmd = "all \"Repeat: Off\"";
                repeatIcon = "sync-ignored"
                break;
        }

        richTooltip.appendMarkdown(this.richTooltipControls.replace("$playCmd",     playing ? "pause" : "play")
                                                           .replace("$playIcon",    playing ? "debug-pause" : "play")
                                                           .replace("$repeatCmd",   repeatCmd)
                                                           .replace("$repeatIcon",  repeatIcon)
                                                           .replace("$shuffleCmd",  shuffle ? "off \"Turn shuffle off\"" : "on \"Turn shuffle on\"")
                                                           .replace("$shuffleIcon", shuffle ? "git-compare" : "checklist"));

        this.titleBarItem.text    = displayedTitle;
        this.titleBarItem.tooltip = richTooltip;
        this.titleBarItem.show();
    }

    public open(): void {
        this.iTunes.open();
    }

    public play(): void {
        this.iTunes.play();
    }

    public pause(): void {
        this.iTunes.pause();
    }

    public nextTrack(): void {
        this.iTunes.nextTrack();
    }

    public previousTrack(): void {
        this.iTunes.previousTrack();
    }

    public shuffleOn(): void {
        this.iTunes.shuffle( true );
    }

    public shuffleOff(): void {
        this.iTunes.shuffle( false );
    }

    public addTrack(): void {
        this.iTunes.addTrack();
    }

    public setRepeat( repeat: string ): void {
        this.iTunes.setRepeat( repeat );
    }

    public dispose(): void {
        this.titleBarItem.dispose();
        clearInterval( this.updateInterval );
    }

    private showMediaControls(): void {
        if (Config.Instance.getShowMediaControls()) {
            this.previousTrackButton.show();
            this.playerButton.show();
            this.nextTrackButton.show();
        }
        if (Config.Instance.getShowLoopShuffle()) {
            this.repeatButton.show();
            this.shuffleButton.show();
        }
        if (Config.Instance.getShowTrack()) {
            this.titleBarItem.show();
        }
        if (Config.Instance.getShowMute()) {
            this.stateButton.show();
        }
        if (Config.Instance.getShowLikeDislike()) {
            this.likeButton.show();
            this.dislikeButton.show();
        }
        if (Config.Instance.getShowAdd()) {
            this.addToLibrayButton.show();
        }
    }

    private hideMediaControls(): void {
        this.previousTrackButton.hide();
        this.playerButton.hide();
        this.nextTrackButton.hide();
        this.repeatButton.hide();
        this.titleBarItem.hide();
        this.stateButton.hide();
        this.shuffleButton.hide();
        this.likeButton.hide();
        this.dislikeButton.hide();
        this.addToLibrayButton.hide();
    }

    public volume(): void {
        if( this.stateButton.text === "$(mute)"){
            this.iTunes.unmute();
        }else{
            this.iTunes.mute();
        }
    }

    public likeTrack(): void {
        this.iTunes.like();
    }

    public dislikeTrack(): void {
        if( Config.Instance.getSkipTrackDislike() === true ) {
            this.iTunes.dislikeTrackAndSkip();
        } else {
            this.iTunes.dislikeTrack();
        }
    }

    public forceUpdate(): void {
        this.onUpdate();
    }
}
